src/Utilities/DoubleDicts.jl

```diff
@@ -132,445 +122,249 @@ function Base.length(d::AbstractDoubleDict)
     return len
 end
 
-function Base.haskey(dict::AbstractDoubleDict, key::CI{F,S}) where {F,S}
-    inner = get(dict.dict, (F, S), nothing)
-    return inner !== nothing ? haskey(inner, key.value) : false
+Base.length(d::AbstractDoubleDictInner) = length(d.inner)
+
+# Base.haskey
+
+function Base.haskey(
+    d::AbstractDoubleDict,
+    key::MOI.ConstraintIndex{F,S},
+) where {F,S}
+    return haskey(getindex(d, F, S), key)
+end
+
+function Base.haskey(
+    d::AbstractDoubleDictInner{F,S},
+    key::MOI.ConstraintIndex{F,S},
+) where {F,S}
+    return haskey(d.inner, key.value)
 end
 
-function Base.get(dict::AbstractDoubleDict, key::CI{F,S}, default) where {F,S}
-    inner = get(dict.dict, (F, S), nothing)
-    if inner !== nothing && haskey(inner, key.value)
-        return typed_value(dict, inner[key.value], F, S)
+# Base.get
+
+function Base.get(
+    d::AbstractDoubleDict,
+    key::MOI.ConstraintIndex{F,S},
+    default,
+) where {F,S}
+    inner = getindex(d, F, S)
+    if haskey(inner, key)
+        return inner[key]
```

Comment:
Cant we avoid haskey+getindex?
This double access pattern always comes back to us with penalties...

Reply:
I've re-organized things, but in this case we do need it.

If we used `get(d.inner, key.value, default)`, we'd then try to `typed_value` map the result back, which wouldn't be correct if the default was returned. And we wouldn't know whether the default was returned because the key was missing, or if the value was the same as the default.